### PR TITLE
Add DateImmutable type

### PR DESCRIPTION
**Pull request by malarzm**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | #1975

#### Summary

As per title :)

### CODE REVIEW

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 1]
```diff
@@ -211,11 +212,11 @@ public function executeInserts(array $options = []) : void
             if ($this->class->isVersioned) {
                 $versionMapping = $this->class->fieldMappings[$this->class->versionField];
                 $nextVersion    = null;
-                if ($versionMapping['type'] === 'int') {
+                if ($versionMapping['type'] === Type::INT || $versionMapping['type'] === Type::INTEGER) {
```
Comment (malarzm) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: I'm not sure why we haven't used consts in the first place so I've used them everywhere and kept it as a separate commit

Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: We should deprecate and remove the `boolean` and `integer` types to be consistent with PHP, which only allows `bool` and `int`. For now this works. 👍

File: lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Doctrine\ODM\MongoDB\Types;
+
+use DateTime;
+use DateTimeImmutable;
+use DateTimeInterface;
+
+class DateImmutableType extends DateType
+{
+    public static function getDateTime($value) : DateTimeInterface
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]:
```suggestion
    /** @return DateTimeImmutable */
    public static function getDateTime($value) : DateTimeInterface
```

File: lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        $datetime = parent::getDateTime($value);
+
+        return $datetime instanceof DateTime ? DateTimeImmutable::createFromMutable($datetime) : $datetime;
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]: This can lead to returning something that isn't a `DateTimeImmutable` if so given from the parent class. I would suggest checking for that.

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php [hunk 4]
```diff
@@ -79,14 +79,14 @@ public function testMultipleFlushesDoIncrementalUpdates()
             $this->dm->persist($test);
             $this->dm->flush();
 
-            $this->assertIsInt($test->getVersion());
+            $this->assertInternalType('int', $test->getVersion());
```
Comment (alcaeus) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php]: Why the change? This method is deprecated in PHPUnit 8.

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php (hunk 1 above)
Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]: Created https://github.com/doctrine/mongodb-odm/issues/2124

File: lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php (hunk 3 above)
Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]: `DateTimeInterface` has only two implementations and IIRC it's forbidden to write a new one in userland. But if you prefer so I can check what's returned and throw an exception if it's neither `DateTime` or `DateTimeImmutable` 👍

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php (hunk 4 above)
Comment (malarzm) [reply tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php]: Somehow I had newest PHPUnit running and it complained while I was running entire `LockTest` class. Can revert it and keep the change for the upgrade PR 👍

File: lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php (hunk 2 above)
Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]: Shall I make similar change to the `DateType`?

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php (hunk 4 above)
Comment (alcaeus) [reply tests/Doctrine/ODM/MongoDB/Tests/Functional/LockTest.php]: Yes please, let's refactor that separately.

File: lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php (hunk 2 above)
Comment (alcaeus) [reply lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]: Yeah, that should help static inspection

Comment (malarzm) [reply lib/Doctrine/ODM/MongoDB/Types/DateImmutableType.php]: FTR: I'm leaving the original DateType as if it detects DateTimeInterface it's returning it right away and DateTime::createFromImmutable was introduced in PHP 7.3 which leaves me with no convenient way of transforming immutable datetime
